zerver/worker/queue_processors.py

```diff
@@ -527,3 +534,43 @@ def consume(self, event: Mapping[str, Any]) -> None:
                 (stream, recipient, sub) = access_stream_by_id(user_profile, stream_id,
                                                                require_active=False)
                 do_mark_stream_messages_as_read(user_profile, stream)
+
+        elif event['type'] == 'export_user_messages':
+            user_profile = get_user_profile_by_id(event['user_profile_id'])
+
+            # Create export
+            output_dir = tempfile.mkdtemp(prefix="/tmp/zulip-export-")
+            do_export_user(user_profile, output_dir)
+            print("Finished exporting to %s; tarring" % (output_dir,))
+            tarball_path = output_dir.rstrip('/') + '.tar.gz'
+            subprocess.check_call(
+                ["tar", "-czf", tarball_path, os.path.basename(output_dir)],
+                cwd=os.path.dirname(output_dir)
+            )
+            print("Tarball written to %s" % (tarball_path,))
+
+            # Upload
+            with open(tarball_path, 'rb') as f:
+                content_type = mimetypes.guess_type(tarball_path)[0]
+                size = os.path.getsize(tarball_path)
+                name = os.path.basename(tarball_path)
+                print(content_type, name, size)
+                upload_path = upload_message_file(name, size, content_type, f.read(), user_profile)
+
+            # Delete export data
+            shutil.rmtree(output_dir)
+            os.remove(tarball_path)
+
+            # Notify user
+            content = """Your data has been exported and uploaded [here]({})""".format(upload_path)
+            internal_send_private_message(user_profile.realm,
+                                          get_system_bot(settings.NOTIFICATION_BOT),
+                                          user_profile,
+                                          content)
```

Comment:
This is a pretty big block of code, which we try to keep out of the raw queue processors; I think it's be better to put this function in `zerver/lib/export.py`.

It might also be a good idea to do a bit of de-duplication with the existing `manage.py` code path by moving code around.

Reply:
Yes, that makes sense. I was a bit wary of doing that in a code base I'm not very familiar with, but I'll give it a shot. Also, since I plan to write some tests, I think it makes sense to do that.